power_grid_model_c/power_grid_model/include/power_grid_model/grouped_index_vector.hpp

```diff
@@ -0,0 +1,148 @@
+        return boost::iterator_range<IdxCount>(indptr_[group], indptr_[group + 1]);
+    }
+    auto get_group(Idx element) const {
+        assert(element < element_size());
+        return static_cast<Idx>(std::upper_bound(indptr_.begin(), indptr_.end(), element) - indptr_.begin() - 1);
+    }
+
+  private:
+    IdxVector indptr_;
+};
+
+class DenseIdxVector {
+  private:
+    class GroupIterator : public boost::iterator_facade<GroupIterator, Idx, boost::random_access_traversal_tag,
+                                                        boost::iterator_range<IdxCount>, Idx> {
+      public:
+        explicit constexpr GroupIterator(IdxVector const& dense_vector, Idx const& group)
+            : dense_vector_{dense_vector},
+              group_{group},
+              group_range_{std::equal_range(std::cbegin(dense_vector_), std::cend(dense_vector_), group)} {}
+
+      private:
+        using group_iterator = IdxVector::const_iterator;
+
+        IdxVector const& dense_vector_;
+        Idx group_;
+        std::pair<group_iterator, group_iterator> group_range_;
+
+        friend class boost::iterator_core_access;
+
+        boost::iterator_range<IdxCount> dereference() const {
+            return boost::counting_range(static_cast<Idx>(std::distance(std::cbegin(dense_vector_), group_range_.first)),
+                                         static_cast<Idx>(std::distance(std::cbegin(dense_vector_), group_range_.second)));
+        }
+        constexpr bool equal(GroupIterator const& other) const { return group_ == other.group_; }
+        constexpr void increment() { advance(1); }
+        constexpr void decrement() { advance(-1); }
+        constexpr void advance(Idx n) {
+            auto const start = n > 0 ? group_range_.second : std::cbegin(dense_vector_);
+            auto const stop = n > 0 ? std::cend(dense_vector_) : group_range_.first;
+
+            group_ += n;
+            group_range_ = std::equal_range(start, stop, group_);
+        }
+    };
+
+    constexpr auto group_iterator(Idx group) const { return GroupIterator{dense_vector_, group}; }
+
+  public:
+    DenseIdxVector() = default;
+    explicit DenseIdxVector(IdxVector dense_vector, Idx groups_size)
+        : dense_vector_(std::move(dense_vector)), groups_size_(groups_size) {}
+
+    constexpr auto size() const { return groups_size_; }
+    constexpr auto begin() const { return group_iterator(Idx{}); }
+    constexpr auto end() const { return group_iterator(size()); }
+
+    constexpr auto element_size() const { return dense_vector_.size(); }
+    constexpr auto get_group(Idx element) const { return dense_vector_[element]; }
+    auto get_element_range(Idx group) const { return *group_iterator(group); }
+
+  private:
+    IdxVector dense_vector_;
+    Idx groups_size_;
+};
+
+template <typename T>
+concept grouped_idx_vector_type =
+    std::is_same<T, const SparseIdxVector>::value || std::is_same<T, const DenseIdxVector>::value;
```

Comment:
shorter + it's good practice to have concepts require other concepts instead of their respective template struct test

```suggestion
concept grouped_idx_vector_type =
    std::same_as<T, const SparseIdxVector> || std::same_as<T, const DenseIdxVector>;
```